### PR TITLE
Add DataProvider interface and impl

### PR DESCRIPTION
**Pull request by jikunshang**

### CODE REVIEW

**Comment by jikunshang** [comment]
@xieqi @yeyuqiang @winningsix  Please take a review. Thanks.

**Comment by winningsix** [comment]
LGTM

**Comment by winningsix** [comment]
Updated since Qi's PR touched this PR.

File: Tests/IterativeExecutionTest.cpp [hunk 1]
```diff
@@ -64,13 +64,16 @@ inline void run_ddl_statement(const std::string& input_str) {
 
 TargetValue run_simple_agg_itr(const std::string& query_str,
                                const ExecutorDeviceType dt) {
+  // it's very tricky since we didn't know the database ID
```
Comment (winningsix) [inline Tests/IterativeExecutionTest.cpp:67]: Caller should know the database ID in the end as they send in the database info.

**Review by winningsix — APPROVED** [review]

**Pull request merged**